### PR TITLE
Fix support for NetBSD and enable native code on NetBSD/aarch64

### PR DESCRIPTION
**Pull request by kit-ty-kate**
Same as https://github.com/ocaml/ocaml/pull/11092 and https://github.com/ocaml/ocaml/pull/11096 but with some fixes to make it actually build.

Was failing with:
```
gcc -c -O2 -fno-strict-aliasing -fwrapv -pthread -Wall -Werror -fno-common -fexcess-precision=standard -ffunction-sections -g -fPIC  -D_FILE_OFFSET_BITS=64  -DCAMLDLLIMPORT=  -o unix.bpic.o unix.c
unix.c: In function 'caml_thread_setname':
unix.c:452:3: error: too few arguments to function 'pthread_setname_np'
  452 |   pthread_setname_np(self, name);
      |   ^~~~~~~~~~~~~~~~~~
In file included from unix.c:52:
/usr/include/pthread.h:160:5: note: declared here
  160 | int pthread_setname_np(pthread_t, const char *, void *);
      |     ^~~~~~~~~~~~~~~~~~
unix.c: In function 'caml_thread_setname':
unix.c:452:3: error: too few arguments to function 'pthread_setname_np'
  452 |   pthread_setname_np(self, name);
      |   ^~~~~~~~~~~~~~~~~~
In file included from unix.c:52:
/usr/include/pthread.h:160:5: note: declared here
  160 | int pthread_setname_np(pthread_t, const char *, void *);
      |     ^~~~~~~~~~~~~~~~~~
unix.c: In function 'caml_thread_setname':
unix.c: In function 'caml_thread_setname':
unix.c:452:3: error: too few arguments to function 'pthread_setname_np'
  452 |   pthread_setname_np(self, name);
      |   ^~~~~~~~~~~~~~~~~~
In file included from unix.c:52:
/usr/include/pthread.h:160:5: note: declared here
  160 | int pthread_setname_np(pthread_t, const char *, void *);
      |     ^~~~~~~~~~~~~~~~~~
unix.c:452:3: error: too few arguments to function 'pthread_setname_np'
  452 |   pthread_setname_np(self, name);
      |   ^~~~~~~~~~~~~~~~~~
In file included from unix.c:52:
/usr/include/pthread.h:160:5: note: declared here
  160 | int pthread_setname_np(pthread_t, const char *, void *);
      |     ^~~~~~~~~~~~~~~~~~
gmake[2]: *** [Makefile:423: unix.b.o] Error 1
```
and
```
gcc -c -O2 -fno-strict-aliasing -fwrapv -pthread -Wall -Werror -fno-common -fexcess-precision=standard -ffunction-sections -g  -D_FILE_OFFSET_BITS=64  -DCAMLDLLIMPORT= -DNATIVE_CODE -DTARGET_arm64 -DMODEL_default -DSYS_netbsd  -o sak.o sak.c
In file included from /usr/include/ctype.h:100,
                 from sak.c:29:
sak.c: In function 'add_stdlib_prefix':
sak.c:126:26: error: array subscript has type 'char' [-Werror=char-subscripts]
  126 |       *name = toupper_os(*name);
      |                          ^
sak.c:126:15: note: in expansion of macro 'toupper_os'
  126 |       *name = toupper_os(*name);
      |               ^~~~~~~~~~
cc1: all warnings being treated as errors
gmake[2]: *** [Makefile:313: sak.o] Error 1
```
All the manual pages necessary to understand those fixes have been added as comments in the code where they are needed.

### CODE REVIEW

File: runtime/sak.c
Comment (avsm) [inline runtime/sak.c]: Given this from that manpage:
```
Some implementations of libc, such as glibc as of 2018, attempt to avoid
     the worst of the undefined behavior by defining the functions to work for
     all integer inputs representable by either unsigned char or char, and
     suppress the warning.  However, this is not an excuse for avoiding con-
     version to unsigned char: if EOF coincides with any such value, as it
     does when it is -1 on platforms with signed char, programs that pass char
     will still necessarily confuse the classification and mapping of EOF with
     the classification and mapping of some non-EOF inputs.
```

Should we just unconditionally cast to `(unsigned int)` on all platforms, including glibc?

Comment (kit-ty-kate) [reply runtime/sak.c]: oooops, I meant to use `unsigned char` not `unsigned int`. Fixed.

Comment (avsm) [reply runtime/sak.c]: Thanks; looking at the glibc man page for toupper, it looks appropriate to always cast `toupper_os()`'s argument to `unsigned char`:

```
The standards require that the argument c for these functions is
       either EOF or a value that is representable in the type unsigned
       char.  If the argument c is of type char, it must be cast to
       unsigned char, as in the following example:

           char c;
           ...
           res = toupper((unsigned char) c);

       This is necessary because char may be the equivalent signed char,
       in which case a byte where the top bit is set would be sign
       extended when converting to int, yielding a value that is outside
       the range of unsigned char.
```

Comment (kit-ty-kate) [reply runtime/sak.c]: Faire enough. Done

**Review by avsm — APPROVED** [review]
The NetBSD fixes look good to me. If I could get another approval on the `toupper` cast before merging, that'd be nice...

File: runtime/sak.c
Comment (Octachron) [inline runtime/sak.c:39]: Am I reading correctly that this change is mainly to avoid a compiler warning on NetBSD?

Comment (kit-ty-kate) [reply runtime/sak.c:39]: Yes. It is triggered by `-Werror=char-subscripts`

Comment (Octachron) [reply runtime/sak.c:39]: Then the `toupper_os` change seems fine to me (and this is the only change that will survive rebasing).

**Comment by Octachron** [comment]
This looks fine to me, I will approve and merge once rebased.

**Comment by Octachron** [comment]
You may want to edit the commit message of the last commit.

File: Changes
Comment (Octachron) [inline Changes]: @avsm should be mentioned as reviewer.

Comment (kit-ty-kate) [reply Changes]: Done

**Review by Octachron — APPROVED** [review]

**Comment by dra27** [comment]
I was just cherry-picking #11639 to 5.0 and it unexpectedly conflicted because this PR isn't there.

I think this PR is intended for 5.0 - is it OK for me to cherry-pick it with #11639?

**Comment by Octachron** [comment]
The PR should have been cherry-picked to 5.0 indeed. So it is fine to cherry-pick it along with #11639 .

**Comment by dra27** [comment]
Done: f35d5d6442

**Pull request merged**
